src/io.c

```diff
@@ -0,0 +1,100 @@
+ *   Ledger App Solana.
+ *   (c) 2023 Ledger SAS.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *****************************************************************************/
+
+#include <stdint.h>
+#include <string.h>
+
+#include "os.h"
+#include "ux.h"
+
+#include "apdu.h"
+#include "menu.h"
+#include "io.h"
+
+// override point, but nothing more to do
+void io_seproxyhal_display(const bagl_element_t *element) {
+    io_seproxyhal_display_default((bagl_element_t *) element);
+}
+
+uint8_t io_event(uint8_t channel) {
+    (void) channel;
+
+    switch (G_io_seproxyhal_spi_buffer[0]) {
+        case SEPROXYHAL_TAG_BUTTON_PUSH_EVENT:
+            UX_BUTTON_PUSH_EVENT(G_io_seproxyhal_spi_buffer);
+            break;
+        case SEPROXYHAL_TAG_STATUS_EVENT:
+            if (G_io_apdu_media == IO_APDU_MEDIA_USB_HID &&  //
+                !(U4BE(G_io_seproxyhal_spi_buffer, 3) &      //
+                  SEPROXYHAL_TAG_STATUS_EVENT_FLAG_USB_POWERED)) {
+                THROW(ApduReplySdkExceptionIoReset);
+            }
+            /* fallthrough */
+        case SEPROXYHAL_TAG_DISPLAY_PROCESSED_EVENT:
+            UX_DISPLAYED_EVENT({});
+            break;
+        case SEPROXYHAL_TAG_TICKER_EVENT:
+            UX_TICKER_EVENT(G_io_seproxyhal_spi_buffer, {});
+            break;
+        default:
+            UX_DEFAULT_EVENT();
+            break;
+    }
+
+    if (!io_seproxyhal_spi_is_status_sent()) {
+        io_seproxyhal_general_status();
+    }
+
+    return 1;
+}
+
+uint16_t io_exchange_al(uint8_t channel, uint16_t tx_len) {
+    switch (channel & ~(IO_FLAGS)) {
+        case CHANNEL_KEYBOARD:
+            break;
+
+        // multiplexed io exchange over a SPI channel and
+        // TLV encapsulated protocol
+        case CHANNEL_SPI:
+            if (tx_len) {
+                io_seproxyhal_spi_send(G_io_apdu_buffer, tx_len);
+
+                if (channel & IO_RESET_AFTER_REPLIED) {
+                    reset();
+                }
+                return 0;  // nothing received from the master so far
+                           // (it's a tx transaction)
+            } else {
+                return io_seproxyhal_spi_recv(G_io_apdu_buffer, sizeof(G_io_apdu_buffer), 0);
+            }
+
+        default:
+            THROW(ApduReplySdkInvalidParameter);
+    }
+    return 0;
+}
+
+void sendResponse(uint8_t tx, bool approve, bool display_menu) {
+    G_io_apdu_buffer[tx++] = approve ? 0x90 : 0x69;
```

Comment:
No defines for these codes ?

Reply:
I would need to add `write_u16_be` and the likes and rework the call to have a proper function behavior, I'll let the devs handle this as it's not related to UI